site/docs/skylark/deploying.md

```diff
@@ -63,6 +62,24 @@ We would have the following structure:
     test.mocs
 ```
 
+### WORKSPACE
+
+In the project's `WORKSPACE`, you should define the name that users will use
+to reference you rules. If your rules belong to the
+[bazelbuild](https://github.com/bazelbuild) organization, you must use
```

Comment:
This conflicts with the documentation here: https://docs.bazel.build/versions/0.28.0/skylark/lib/globals.html#workspace

Do we really need to make a special-case for `bazelbuild`? Is it a problem if we always use `rules_<lang>`, no matter what is the organization?

(cc @aehlig)

Reply:
Regarding the conflict: I haven't found any project in the bazelbuild org that actually follows this convention of `com_github_bazelbuild_rules_<lang>`. Some are named `io_bazel_rules_{closure,go}`, some are named `build_bazel_rules_apple`, and some are named `rules_{cc,java}`. Anyways, we should definitely bring them in sync and add a link to the style guide for rules.

Always using `rules_<lang>` does not seem like a good idea because it will lead to independent, hence incompatible, repos with the same name (https://awesomebazel.com list at least two `rules_{scala,terraform,proto,...}`). Let's assume we have the `rules_<lang>` rule and everyone followed it: Once there was https://github.com/pubref/rules_proto (`pubref`) which got deprecated in favor of https://github.com/stackb/rules_proto (`stackb`). If any project in the transitive closure of dependencies depends on `pubref`, no one in the transitive closure can depend on `stackb`. It should be obvious that that's a problem and that it would be better for the Bazel ecosystem if everyone follows `<org>_rules_<lang>`.

The exception for `bazelbuild` is for repos like `bazelbuild/rules_{cc,java,proto,...}` that pretty much every single user of Bazel will depend on, so they should have a short and memorable name. If we decide against a special-case for `bazelbuild`, I think these should follow `<org>_rules_<lang>` as well.

Reply:
> This conflicts with the documentation here: https://docs.bazel.build/versions/0.28.0/skylark/lib/globals.html#workspace
> 
> Do we really need to make a special-case for `bazelbuild`? Is it a problem if we always use `rules_<lang>`, no matter what is the organization?
> 
> (cc @aehlig)

Special casing adds more complexity to an already complex topic. And yes, the fact that `bazelbuild` doesn't even follow its own naming convention doesn't help either.

However, the more fundamental problem is that this more than just informative what the repository is about. This name is used in many rules repositories to refer to targets in that repository. While the [proposals repository](https://github.com/bazelbuild/proposals/) refers to a [document on label expansion](https://docs.google.com/document/d/1bEhKy1lQB-EIpNxSYgRLtZ8p07dRe0g-t7D5j5jxLb0/) that in theory should make it unnecessary for rules repositories to know their own name, that doesn't seem to work in practice. Macros passing targets around as strings make consistent expansion even harder.

Now, if repositories need to know their own name, it better be unique to avoid clashes&mdash;and that's where the recommendation of fully-qualified names came from. While we have a mechanism to rename repositories, that, of course, doesn't work if targets are passed around as strings rather than as labels.

Reply:
We really have to fix that issue - that repositories have to know their own  name. I keep running into it trying to write rules that test locally and also work when either vendored into  or referenced in an another repo.  We've been lingering on a partial solution for 6 months now.
#7130  but can't seem to come to closure on it.

Does the label expansion proposal (
https://docs.google.com/document/d/1bEhKy1lQB-EIpNxSYgRLtZ8p07dRe0g-t7D5j5jxLb0/edit) play well with the change for #7130 ?
Do we need to update it to talk about the "magic" repositories, like @bazel_tools

Reply:
I likewise challenge the stated rationale for special casing bazelbuild. If Bazel becomes a healthy ecosystem at all, we can expect very commonly used rules outside of that organization, maintained by third parties. @Yannic says that we should have unqualified names for common rules because they're common. But,

1. most users needing e.g. `rules_cc` at least once doesn't mean that `rules_cc` is the most used rule set for most users, and
2. if long names are such a problem that they're unbearable for widely used rule sets, why should they be fine for less commonly used ones? They're still very common for each of their users.

To me, that long domain qualified names is a problem for some rules shows that it's a problem for all rules and it would be nice to avoid this problem. Especially if we're not going to have a BUILD file prelude (see https://github.com/bazelbuild/bazel/issues/3835 and https://github.com/bazelbuild/bazel/issues/1674).